source/async_stack.cpp

```diff
@@ -39,10 +42,47 @@ extern "C" {
 inline pthread_key_t folly_async_stack_root_tls_key = 0xFFFF'FFFFu;
 }
 
-#endif  // FOLLY_ASYNC_STACK_ROOT_USE_PTHREAD
+#else
+
+#include <vector>
+
+// TODO: expose array of AsyncStackRootHolder
+
+#endif  // ASYNC_STACK_ROOT_USE_PTHREAD
 
 namespace unifex {
 
+#if ASYNC_STACK_ROOT_USE_PTHREAD == 0
+static std::unique_ptr<std::mutex> asyncStackRootsMutex = std::make_unique<std::mutex>();
+static std::vector<void*> asyncStackRoots;
+
+static void addAsyncStackRootHolder(void* holder) noexcept {
+  std::lock_guard<std::mutex> lock(*asyncStackRootsMutex);
+  asyncStackRoots.push_back(holder);
+}
+
+static void removeAsyncStackRootHolder(void* holder) noexcept {
+  if (asyncStackRootsMutex == nullptr) {
+    return;
+  }
+  std::lock_guard<std::mutex> lock(*asyncStackRootsMutex);
+  auto it = std::find(asyncStackRoots.begin(), asyncStackRoots.end(), holder);
+  if (it != asyncStackRoots.end()) {
+    asyncStackRoots.erase(it);
+  }
+}
+
+// TODO: expose array of AsyncStackRootHolder
+// static std::vector<void*> getAsyncStackRoots() noexcept {
+//   if (!asyncStackRootsMutex.try_lock()) {
+//     // assume we crashed holding the lock and give up
+//     return {};
+//   }
+//   std::lock_guard<std::mutex> lock(asyncStackRootsMutex, std::adopt_lock);
+//   return asyncStackRoots;
+// }
```

Comment:
Maybe you should expose this as a type. Something like:

```cpp
struct AsyncStackRootHolderList {
  std::vector<void*> holders_;
  std::mutex mutex_;

  void add(void*) { … }

  void remove(void*) { … }
};

extern "C" auto* kUnifexAsyncStackRootHolderList = new AsyncStackRootHolderList();
```

If you leak it like that, the main thread won't destroy it. You can also find the start of the vector from out of process by looking up the symbol. Not sure how you make a copy of the vector from out of process…